src/components/PassengerShowcaseForm.js

```diff
@@ -56,11 +54,92 @@ const PassengerShowcaseForm = () => {
           videoNumber
         }
       }
+      tracks: allTrack {
+        edges {
+          node {
+            title
+            slug
+            chapters {
+              title
+              videos {
+                title
+                slug
+                canContribute
+                canonicalTrack {
+                  slug
+                }
+              }
+            }
+            videos {
+              title
+              slug
+              canContribute
+              canonicalTrack {
+                slug
+              }
+            }
+          }
+        }
+      }
     }
   `);
 
+  const tracks = data.tracks.edges
+    .map((edge) => edge.node)
+    .map((track) => {
+      // gather all videos from chapters (main track)
+      if (!track.videos && track.chapters) {
+        track.videos = track.chapters.reduce((acc, chapter) => {
+          return acc.concat(chapter.videos);
+        }, []);
+      }
+      delete track.chapters;
+      return track;
+    })
+    .map((track) => {
+      // filter out videos that can't be contributed to
+      track.videos = track.videos.filter((video) => video.canContribute);
+      // keep only videos that belong to this track
+      track.videos = track.videos.filter(
+        (video) => video.canonicalTrack?.slug === track.slug
+      );
+      return track;
+    })
+    .filter((track) => track.videos.length > 0);
+
+  // create a "challenges track"
+  const challengesTrack = {
+    title: 'Coding Challenges',
+    slug: 'challenges',
+    videos: data.challenges.nodes.map((node) => {
+      return {
+        title: `#${node.videoNumber} ${node.title}`,
+        slug: node.slug
+      };
+    })
+  };
+  return [challengesTrack, ...tracks];
```

Comment:
It would be a nice optimization to memoize this whole transform and filtering operation. The hook will run every time a form value is updated yet always returns the same value.

Reply:
I thought about this but couldn't find a good way to do this. Is it maybe a good idea to save the output in a state variable on mount?

Reply:
Yeah, useEffect + useState could work, but memo would be even simpler. Inside the hook itself, try something like this:

```js
const data = useStaticQuery // ...

return useMemo(()=>{ /* your filter code */ }, [data]);
```

Reply:
That makes sense, thanks for your input!